### ISSUE TITLE
(Twoslash) Bug: Two files in a single codeblock can fail

### ISSUE THREAD

**Issue by juancampa** [MatthiesenXYZ/EC-Plugins#36]
**Describe the bug**

When using the twoslash plugin, if you have a codeblock with multiple files, for example:

```
// @filename module.ts

...

// @filename index.ts
import * from "module";

...
```

Typescript will fail to compile because it can't find the module named `module`.

After digging into it, I noticed that the first `@filename` was missing by the time the code reaches twoslash. It turns out that the ec "frames" plugin is stripping it away.

The fix is to use this config:

```js
{
  plugins: [ecTwoSlash({})],
  frames: {
    extractFileNameFromCode: false,
  },
},
 ```

It wasn't obvious at all and it makes the plugin feel broken. Should we add it to [these instructions in the docs](https://twoslash.matthiesen.dev/getting-started/installation/)?

**Comment by Adammatthiesen** [comment]
Hi, the `// @filename` syntax is actually part of the [Show Emitted Files](https://twoslash.matthiesen.dev/getting-started/installation/#twoslash-features) system that is part of Twoslash (more specifically the Twoslash VFS) and is not yet supported in EC-Twoslash

Though this information does help me understand part of the reasons things with the Twoslash VFS system aren't working with EC-Twoslash, i actually pull in the full raw code block and overwrite it with the Twoslashed output... so the filename tags are supposed to still be there in the code since i use one of the earliest EC hooks available to plugins to process the whole code block.

A good part of this issue is actually related to the way Twoslash VFS works... or well doesn't work with EC. And have been trying to get a better understanding of what they (Twoslash) are doing under the hood to be able to re-implement this system within EC correctly.

I have an open issue: https://github.com/twoslashes/twoslash/issues/59 but have not gotten any responses.

**Comment by hippotastic** [comment]
If changing the config of the default plugin `plugin-frames` actually fixes this, I'd consider this an EC issue I can fix rather than a Twoslash issue. I could ensure that the special `@filename` syntax is ignored by the frames plugin in order not to interfere with Twoslash.

**Comment by Adammatthiesen** [comment]
> If changing the config of the default plugin `plugin-frames` actually fixes this, I'd consider this an EC issue I can fix rather than a Twoslash issue. I could ensure that the special `@filename` syntax is ignored by the frames plugin in order not to interfere with Twoslash.
> 
> As my default plugins get prepended to the plugins array internally by the `expressive-code` package, it could very well be that the `plugin-frames` processing runs before your Twoslash plugin is able to access the code.

@hippotastic Did you end up changing this?  i was playing around and had it working on the StudioCMS docs without any config changes... 🤔 So not really sure what the current status of this issue is

**Comment by hippotastic** [comment]
No, the code has not been touched, I actually forgot about this thread. Sorry about that!

The reason why it might have worked for you is that you had a title set for the code block, either through props or the meta information. In that case, no attempt will be made to extract a file name for the title.

**Comment by Adammatthiesen** [comment]
> No, the code has not been touched, I actually forgot about this thread. Sorry about that!
> 
> The reason why it might have worked for you is that you had a title set for the code block, either through props or the meta information. In that case, no attempt will be made to extract a file name for the title.

That would be exactly what it looks like 🤔 

https://github.com/withstudiocms/studiocms/blob/e3578734d8d96fbac80dc9941192361ab8b6dbb8/docs/src/content/docs/plugins/extended.mdx?plain=1#L129-L159

**Comment by hippotastic** [comment]
Ok. Can you confirm that it breaks if you remove the title?

**Comment by Adammatthiesen** [comment]
![Image](https://github.com/user-attachments/assets/93877c33-5e37-45af-a4e0-9d4706ed7284)

![Image](https://github.com/user-attachments/assets/bfbaaa51-1147-4928-b1b5-77d46b4eb3c4)

Well it didn't exactly break, just added the FIRST filename to the title... so that was not was i was expecting. 😅 

Maybe its because i use a relative import since its not a module but a regular `.ts` file?

**Comment by Adammatthiesen** [comment]
![Image](https://github.com/user-attachments/assets/4c6d7759-456d-4745-9cdd-09db21064eb0)

Wait im an idiot... i had a `@noErrors` tag, 

- [6053] 274 - File '/home/adam/repos/studiocms/docs/plugin.d.ts' not found.
-  [2307] 486 - Cannot find module 'myplugin:config' or its corresponding type declarations.

**Comment by Adammatthiesen** [comment]
So yes, the `@filename` thing is still a problem 😅 If you dont set a `title=""` attribute

**Comment by hippotastic** [comment]
Ok, thank you! I'll publish a fix then. Anything else other than @filename I should ignore?

**Comment by Adammatthiesen** [comment]
Upstream issue has been created: https://github.com/expressive-code/expressive-code/issues/427
